### DISCUSSION TITLE
simulate central-place foraging trips with simData

### DISCUSSION THREAD

**Discussion by CLambert1** [bmcclintock/momentuHMM#132] [Q&A]
I am currently building a virtual species within a completely virtual environment. I simulated two covariates from which I build a suitability map that will be used by the virtual species. I consider two colonies, and I want to simulate the movements of 2*10 000 individuals with the constraint of central-place foraging. For now, using the simData function, I was able to simulate tracks constrained on the suitability and distance from the colony, but I would like to also force the tracks to go back to the colony; that is I would like the last step of the track to be at the starting point. The aim is to mimic seabird central-place behaviour and simulate a full trip for each individual. How could I achieve this with the simData function? 
Thanks for your kind help !
Cheers
Charlotte

**Comment by bmcclintock** [comment]
Using the discrete-time step-and-turn approach, you could incorporate centers of attraction via the `centers` argument in `simData` as demonstrated in the grey seal and northern fulmar examples in the `momentuHMM` [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf). You would likely need to make the state transition probabilities a function of distance to the center of attraction and further constrain the state transitions so that they follow the appropriate order (e.g. "outbound" -> "foraging" -> "inbound" similar to the southern elephant seal example in the [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf)). The only way to guarantee the tracks end at the center of attraction would be to make the "inbound" state absorbing (i.e. so they can't transition out of the "inbound" state) and make sure that the lengths of the tracks are sufficiently long to allow them to return to the center of attraction once they are in the "inbound" state. You could instead just make transitions out of the "inbound" state a function of distance to the center of attraction (such that they can't transition until they reach the center of attraction) and then just truncate the simulated tracks at the point they returned to the center of attraction.

Alternatively, you could model the positions directly and use a potential function as demonstrated in #111 for a hypothetical central-place forager.

**Comment by CLambert1** [comment]
Thanks for showing the way to go. I managed to adapt the code from the southern elephant seal example, but there are some points I quite don't grasp yet. 
How should I code the formulas to condition angles and transition probability on angle and distance from the center of attraction for some states? I understand how to do so when fitting a model on an existing track (values are associated to existing points so it is straightforward), but not when you simulate from scratch (which I do). Similarly, I would also like to impose an effect on time (number of steps) on the transition probability. 

For now on I have something like this, which doesn't work (at the very least I don't have the angles right), where I condition step length and turning angle to the distance from colony (input is a raster layer), and transition probability to distance from colony and underlying environment:

```
nbStates <- 4
stateNames <- c("outbound", "search", "forage", "inbound")

formula <- ~ betaCol1(dist_col) + # transition from outbound to forage depend on distance from colony
                      betaCol6(-dist_col) +  # transition from search to inbound depend on distance to colony
                      betaCol5(env)         # transition from search to forage depend on the underlying environment
                      #+ betaCol6(time)   # I would like the transition to inbound state to depend on time (number of step)
distFormula <- ~ state1(dist_col) + state4(-dist_col)
stepDM <- list(mean = distFormula, sd = distFormula)
angleDM <- list(mean = distFormula, concentration = distFormula)

fixbeta <- matrix(c(NA,-100,-100,-100,NA,NA,-100,NA,-100,-100,-100,-100, # intercept
                    rep(NA,12), # dist_col; will be put automatically to 0 by the function (if I understood correctly the examples)
                    rep(NA,12)), # env
                  nrow=3,byrow=TRUE)
fixPar <- list(angle = c(0,1,3,0, 0.1,2,2,0.1), beta = fixbeta,
               step = c(25,5,1,25, 10,5,3,10)) 

par <- getParDM(DM = list(step = stepDM, angle = angleDM), 
         Par = fixPar, 
         dist = list(step = "gamma", angle = "vm"),
         nbStates = nbStates,
         data = data.frame(dist_col = as.data.frame(dist_col1_scaled), 
                           env = resource$suitability),
         circularAngleMean = list(angle = T),
         estAngleMean = list(angle = T)
  )

sim_track <- simData(nbAnimals = 1, nbStates = nbStates, stateNames = stateNames,
                     obsPerAnimal = 1000, 
                     states = TRUE,
                     retrySims = 100,
                     initialPosition = coord_col1,
                     centers = matrix(coord_col1, nrow = 1, ncol = 2, byrow = TRUE),
                     DM = list(step = stepDM, angle = angleDM), 
                     formula = formula, 
                     Par = par, 
                     dist = list(step = "gamma", angle = "vm"),
                     spatialCovs = list(env = env_rast,
                                        dist_col = dist_col1))
```

Thank you very much again for you help ;)

**Reply by bmcclintock** [reply]
I cannot diagnose your code without a reproducible example, but `simData` internally calculates the angles and distances from centers of attraction specified using the `centers` argument. See `?simData` (specifically the `centers` argument), the gray seal example in the `momentuHMM` [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf), and the corresponding [code](https://github.com/bmcclintock/momentuHMM/blob/master/vignettes/examples/greySealExample.R). If you want to include number of steps as a covariate, then this must be provided as a covariate using the `covs` argument, e.g., `covs = data.frame(time=1:1000)`.

**Comment by CLambert1** [comment]
We reworked a bit our example and incorporated the center of attraction. The characteristics of movements for each state are correctly simulated (large turning angles and low steps for foraging state; long steps and low turning angles for inbound and outbound), but most of/all the steps are simulated to belong to the last state (here, inbound), and we cannot figure out why. Playing with the values of the beta matrix did not change much things; we also tried to add a "colony" state (as in the grey seal example), with the same results. In addition, the outbound and inbound states are not directed from/towards the center of attraction as we would like it to be.  
What did we miss for the simulation to successfully identify the three states? 
Thanks again for your help.
Here is the minimal example: 

```
# Load required packages
library(ggplot2)
library(Rcpp) # for likelihood in C++
library(raster) # to extract covariates
library(ash) # for bin2
library(gstat)
library(momentuHMM)
library(dplyr)
library(sf)
library(sp)

### 1 - Define the grid --------------------------------------------------------
ymin = 0
ymax = 90
xmin = 0
xmax = 90
step.x = 0.2
step.y = 0.2

xy <- expand.grid(seq(xmin, xmax, step.x), seq(xmin, ymax, step.y))
colnames(xy) <- c("x","y")


### 2 - Define and predict two resource distributions --------------------------
set.seed(100)

# set up
field <- gstat(formula=z~1, locations=~x+y, dummy=TRUE, beta=1, 
               model=vgm(psill=100, range=100, model='Exp'), nmax=20)
cov <- predict(field, newdata=xy, nsim=1)

cov[,3] <- (cov[,3]-min(cov[,3]))/(max(cov[,3])-min(cov[,3]))

# transform to rasters
env_rast <- rasterFromXYZ(cov[,c("x","y","sim1")])
names(env_rast) <- "env"
cov1 <- as.matrix(env_rast)


### 4 - Compute distance from colonies -----------------------------------------
coord_col <- c(36,55)

dist_col <- distanceFromPoints(env_rast, xy = coord_col)
names(dist_col) <- "dist_col"
proj4string(dist_col) <- sp::CRS("+init=epsg:4326") 

### 4 - Compute movement ------------------------------------------------------
## define the states
nbStates <- 3
stateNames <- c("outbound", "forage", "inbound")


## define transition probability (beta)
formula <- ~betaCol1(env) + betaCol4(time)
fixbeta <- matrix(c(75,-100,-100,0,-100,-100, # intercept
                    rep(0,6), # env
                    rep(0,6) # time
),
nrow = 3, byrow = TRUE,
dimnames = list(c("Intercept","env","time"),
                c("1 -> 2","1 -> 3",
                  "2 -> 1","2 -> 3",
                  "3 -> 1","3 -> 2")))
# steps are longer and less variable for out and inboud states
# angles are larger and more variable for foraging state
fixPar <- list(angle = c(0,2,0,  0.05,0.2,0.05), # shape and scale (conc)
               step = c(1.5,0.5,1.5,  0.05,0.5,0.05),# mean1, mean2, scale1, scale2
               beta = fixbeta)


## effect of covariates on the step length and turning angle concentration
stepDM <- list(mean = ~ state2(env), sd = ~ 1) 
angleDM <- list(mean = ~ 1, 
                concentration = ~ state2(-env))


## convert natural scale parameters to working scales
par <- getParDM(nbStates = nbStates,
                dist = list(step = "gamma", angle = "vm"),
                Par = fixPar,
                DM = list(step = stepDM, angle = angleDM), 
                data = data.frame(env = as.data.frame(env_rast)[1:500,], 
                                  dist_col = as.data.frame(dist_col)[1:500,],
                                  gradcenter.x = as.data.frame(gradcenter.x)[1:500,],
                                  gradcenter.y = as.data.frame(gradcenter.y)[1:500,],
                                  time = 1:500),
                estAngleMean = list(angle = T)
)

## simulation
sim_track <- simData(nbAnimals = 1, nbStates = nbStates, stateNames = stateNames,
                     obsPerAnimal = 1000, 
                     states = TRUE,
                     retrySims = 10,
                     initialPosition = coord_col,
                     centers = matrix(coord_col, nrow = 1, ncol = 2, byrow = TRUE),
                     dist = list(step = "gamma", angle = "vm"),
                     formula = formula, 
                     Par = list(angle = par$angle, 
                                step = par$step),
                     beta = fixbeta,
                     delta = c(1-2.e-100,1.e-100,1.e-100), # starts in state 1
                     DM = list(step = stepDM, angle = angleDM),
                     covs = data.frame(time=1:500), 
                     spatialCovs = list(env = env_rast,
                                        dist_col = dist_col,
                                        gradcenter.x = gradcenter.x,
                                        gradcenter.y = gradcenter.y ))
ggplot(sim_track) + geom_point(aes(x = x, y = y, color = as.factor(states)))

```

**Comment by bmcclintock** [comment]
Again, I cannot diagnose without a minimal _reproducible_ example, but it appears the covariate names in `formula`, `stepDM`, and `angleDM` do not correspond to those internally generated by the `centers` argument (e.g. `center1.dist`, `center1.angle`). I suggest you look more closely at `?simData` (specifically the `centers` argument), the gray seal example in the momentuHMM [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf), and the corresponding [code](https://github.com/bmcclintock/momentuHMM/blob/master/vignettes/examples/greySealExample.R). Building on the grey seal example code, here is an example of simulating the track "from scratch" without using the `model` argument:
```
# set parameters
Par=list(Par=list(step=c(0.220181463852411, 0.150253194174099, 0.16339565139876, 1.07586885028161, 0.393063758823201, 1.09409883090581, 0.525997528992793, -1.18780598967134, 7.35193980207823, 1.03382421312728, 7.39178473131803, 1.69232802658152, 7.87000828222016, 1.25675673731686, 7.61846840121869, -0.459907614562146),
                  angle=c(1, 1, 1, -2.02070213294872, 2.5744698338948, -0.645371193928587, 1.93236338556476, -1.13803829342907, 2.67311878424043, -100, -100)),
         beta=matrix(c(-8.23984091157603, -0.80610653016325, 0, 0, -7.78133949230777, 3.94555718921247, 0, 0, -8.23907691678559, -0.777446443368415, 0, 0, -8.16574185346988, -0.0751552551516316, 0, 0, -6.48924405408264, 0, 2.5526826048766, 0, -6.72962970476114, 0, 0.470725292501938, 0, -6.79411821261795, 0, 0.0647705352201481, 0, -6.45065080059252, 0, 3.0612846295506, 0, -8.53286542662232, 0, 0, -0.810109507174362, -8.12353127795748, 0, 0, 3.48806094849804, -8.27905573782234, 0, 0, 1.93033182269331, -7.9744178813277, 0, 0, 5.51164021650008, -6.68707731944078, 0, 0, 0, -6.06608514252377, 0, 0, 0, -5.52698682319384, 0, 0, 0, -2.84241459492851, 0, 0, 0, -6.47653468991759, 0, 0, 0, -3.00430253064852, 0, 0, 0, -2.51554221433294, 0, 0, 0, -3.07293948512567, 0, 0, 0),4),
         delta=c(0.99998147075563, 3.94595305770285e-06, 4.86108592340391e-06, 4.86110633074871e-06, 4.86109905808838e-06))

# simulate track
setRNG::setRNG(kind="L'Ecuyer-CMRG",normal.kind="Inversion",seed=2484)
greySealSim<-simData(nbStates=nbStates,dist=dist,
                     Par=Par$Par,beta=Par$beta,delta=Par$delta,
                     circularAngleMean=circularAngleMean,
                     DM=list(step=MIstepDM,angle=angleDM),
                     formula=distFormula,
                     workBounds=list(step=stepworkBounds),
                     stateNames=stateNames, centers=centers, initialPosition = centers[1,], obsPerAnimal = 1515, states=TRUE)

head(greySealSim)
#  ID      step      angle        x       y Abertay.dist Abertay.angle Farne.dist Farne.angle Dogger.dist Dogger.angle states
#1  1  591.8604         NA 514224.9 6254724       0.0000      0.000000   114372.3  -0.8718418    373661.5   -0.3117669      1
#2  1  480.7459  2.8643145 514598.4 6255184     591.8604      3.141593   114484.9  -1.7648168    373447.3   -1.2011411      1
#3  1  530.9646 -0.1699692 514204.5 6254908     184.6146      1.070803   114525.9   1.6582371    373737.2    2.2187561      1
#4  1  924.1771  0.5081910 513724.3 6254681     502.4675      2.786595   114662.2   1.8326844    374124.9    2.3896956      1
#5  1 1102.2964  1.6426094 513185.8 6253930    1307.7740      2.845304   114440.4   1.3323253    374408.5    1.8838547      1
#6  1 3232.9132 -1.4586602 514125.5 6253354    1373.9930      2.048515   113391.2  -0.3132522    373338.2    0.2419507      1

# plot track
colvect<-c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2")
leg=character(5)
leg[1]="Abertay haul-out state"
leg[2]="Farne Islands haul-out state"
leg[3]="Dogger Bank foraging state"
leg[4]="Low-speed exploratory state"
leg[5]="High-speed exploratory state"

plot(greySealSim$x,greySealSim$y,type="o", asp=1, pch=20,ylab="Latitude",xlab="Longitude",ylim=c(6060000,6289000),xlim=c(513000,925000),xaxt="n",yaxt="n",xaxs="i",yaxs="i")
points(greySealSim$x,greySealSim$y,col=colvect[greySealSim$states],pch=20)
axis(1,labels=c(-2,0,2,4),at=c(565312,696098,(434680-312928+696098),(565330-312928+696098)))
axis(2,labels=c(55,56,57),at=c(6094820,6206100,6317500))
legend("topright",leg,pch=rep(20,length(leg)),col=colvect,bty="n")
points(centers,col="#D55E00",pch=20,cex=1.6)
```
![image](https://user-images.githubusercontent.com/11667599/168146528-e3d0b36e-256d-409c-b5a2-867928b2c042.png)

Note that the covariate names in `DM` and `formula` (e.g. `Abertay.dist`, `Dogger.angle`) match those internally generated by `simData` based on the `centers` argument.
